source/common/upstream/upstream_impl.cc

```diff
@@ -2095,7 +2095,24 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(
 
       // If we are depending on a health checker, we initialize to unhealthy.
       if (health_checker_ != nullptr && !host->disableActiveHealthCheck()) {
-        host->healthFlagSet(Host::HealthFlag::FAILED_ACTIVE_HC);
+        if (Runtime::runtimeFeatureEnabled(
+                "envoy.reloadable_features.keep_endpoint_active_hc_status")) {
+          if (existing_host_found && !health_check_address_changed &&
+              !active_health_check_flag_changed) {
+            // If there's an existing host, use the same active health-status.
+            if (existing_host->second->healthFlagGet(Host::HealthFlag::FAILED_ACTIVE_HC)) {
+              host->healthFlagSet(Host::HealthFlag::FAILED_ACTIVE_HC);
+            }
+            if (existing_host->second->healthFlagGet(Host::HealthFlag::DEGRADED_ACTIVE_HC)) {
+              host->healthFlagSet(Host::HealthFlag::DEGRADED_ACTIVE_HC);
```

Comment:
The health-check flags bits also include non-active health check flags, so copying everything over may not make sense here.
I've copied the 4 relevant active health-check flags. Also added a couple of assertions, because some flags can only be set with others.